File: backend/schemas/charts.py
# backend/schemas/charts.py
"""
Pydantic schemas for charts endpoints.
"""
from __future__ import annotations
from typing import Optional, List, Dict, Any
from datetime import datetime

from pydantic import BaseModel, Field


# ============================================================================
# REQUEST SCHEMAS
# ============================================================================

class FollowChartRequest(BaseModel):
    """Request to follow a chart"""
    top_n_artists: int = Field(default=40, ge=1, le=40, description="Number of top artists to auto-follow (1-40)")


class UpdateChartRequest(BaseModel):
    """Request to update chart subscription settings"""
    top_n_artists: Optional[int] = Field(None, ge=1, le=40, description="Number of top artists to auto-follow (1-40)")
    enabled: Optional[bool] = Field(None, description="Enable/disable chart sync")


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================

class ChartArtist(BaseModel):
    """Artist in a chart"""
    id: str
    name: Optional[str] = None
    thumbnails: List[Dict[str, Any]] = Field(default_factory=list)
    rank: Optional[int] = None
    trend: Optional[str] = None  # "up", "down", "same", "new"
    followed: bool = False  # Whether this artist is already followed
    
    model_config = {"from_attributes": True}


class ChartSubscriptionResponse(BaseModel):
    """Chart subscription details"""
    id: int
    country_code: str
    enabled: bool
    top_n_artists: int
    created_at: datetime
    created_by: Optional[int] = None
    last_synced_at: Optional[datetime] = None
    last_error: Optional[str] = None
    
    model_config = {"from_attributes": True}


class ChartResponse(BaseModel):
    """Chart data with subscription status"""
    country_code: str
    artists: List[ChartArtist]
    followed: bool = False
    subscription: Optional[ChartSubscriptionResponse] = None
    cached: bool = False

File: frontend/src/api/charts.ts
// src/api/charts.ts
/**
 * Charts API endpoints
 */

import { api } from './client';

// ============================================================================
// TYPES
// ============================================================================

export interface ChartArtist {
  id: string;
  name: string;
  thumbnails: Array<{
    url: string;
    width: number;
    height: number;
  }>;
  rank: number;
  trend: 'up' | 'down' | 'neutral';
  followed: boolean;
}

export interface Chart {
  country_code: string;
  artists: ChartArtist[];
  followed: boolean;
  subscription: ChartSubscription | null;
  cached: boolean;
}

export interface ChartSubscription {
  id: number;
  country_code: string;
  enabled: boolean;
  top_n_artists: number;
  created_at: string;
  created_by: number;
  last_synced_at: string | null;
  last_error: string | null;
}

export interface FollowChartRequest {
  top_n_artists: number;
}

export interface UpdateChartRequest {
  top_n_artists?: number;
  enabled?: boolean;
}

export interface MessageResponse {
  message: string;
}

// ============================================================================
// CHART ENDPOINTS
// ============================================================================

/**
 * Get chart for a specific country
 */
export async function getChart(countryCode: string): Promise<Chart> {
  return api.get<Chart>(`/charts/${countryCode}`);
}

/**
 * List all chart subscriptions
 */
export async function listChartSubscriptions(includeDisabled = false): Promise<ChartSubscription[]> {
  return api.get<ChartSubscription[]>('/charts', { include_disabled: includeDisabled });
}

/**
 * Follow a chart (admin only)
 */
export async function followChart(countryCode: string, topNArtists: number): Promise<ChartSubscription> {
  return api.post<ChartSubscription>(`/charts/${countryCode}/follow`, { top_n_artists: topNArtists });
}

/**
 * Unfollow a chart (admin only)
 */
export async function unfollowChart(countryCode: string): Promise<MessageResponse> {
  return api.delete<MessageResponse>(`/charts/${countryCode}/follow`);
}

/**
 * Update chart subscription (admin only)
 */
export async function updateChart(
  countryCode: string,
  updates: UpdateChartRequest
): Promise<ChartSubscription> {
  return api.patch<ChartSubscription>(`/charts/${countryCode}`, updates);
}